### PR TITLE
feat(docs): add types grammar to spec

### PR DESCRIPTION
**Pull request by jlapacik**
### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

### CODE REVIEW

File: docs/SPEC.md [hunk 1]
```diff
@@ -405,10 +405,28 @@ The default value is set using the time zone of the running process.
...
 
 A type defines a set of values and operations on those values.
-Types are never explicitly declared as part of the syntax.
+Types are never explicitly declared as part of the syntax except as part of a [builtin statement](#built-ins).
 Types are always inferred from the usage of the value.
 Type inference follows a Hindley-Milner style inference system.
 
+    polytype = monotype ["where" constraints] .
+    monotype = tvar | basic | array | record | function .
+    tvar     = "A" … "Z" .
```
Comment (nathanielc) [inline docs/SPEC.md:414]: do we want to allow for `t#` format types as well in cases where there are lots of type variables? Or only use letters?

File: docs/SPEC.md [hunk 2]
```diff
@@ -405,10 +405,28 @@ The default value is set using the time zone of the running process.
...
+    basic    = "int" | "uint" | "float" | "string" | "bool" | "time" | "duration" | "bytes" | "regexp" .
+    array    = "[" monotype "]" .
+    record   = "{" [properties] "}" | "{" tvar "with" properties "}" .
+    function = "(" [parameters] ")" "->" monotype .
```
Comment (nathanielc) [inline docs/SPEC.md:418]: Should this be `=>` instead of `->`?

Comment (jlapacik) [reply docs/SPEC.md:418]: Yes this is supposed to be `=>`

File: docs/SPEC.md (hunk 1 above)
Comment (jlapacik) [reply docs/SPEC.md:414]: I think we should only use letters for now. To need any more than 3 type variables will be extremely rare at best.

File: docs/SPEC.md [hunk 3]
```diff
@@ -1085,11 +1009,29 @@ These preassigned values are defined in the source files for the various built-i
...
+    Parameter  = [ "<-" | "?" ] identifier ":" MonoType .
+
+    Constraints = Constraint { "," Constraint } .
+    Constraint  = Tvar ":" Kinds .
+    Kinds       = identifier { "+" identifier } .
 
 Example
 
-    builtin from : (bucket: string, bucketID: string) -> stream
+    builtin filter : (<-tables: [T], fn: (r: T) -> bool) -> [T]
```
Comment (nathanielc) [inline docs/SPEC.md:1034]: do we want to include the `where T : Record` bit to this example?

Comment (jlapacik) [reply docs/SPEC.md:1034]: I think I'll leave it out for now just because the constraints themselves aren't part of the spec.

**Pull request merged**